File: library/src/main/java/com/smartmobilefactory/epubreader/display/vertical_content/VerticalEpubWebView.java
package com.smartmobilefactory.epubreader.display.vertical_content;

import android.content.Context;
import android.util.AttributeSet;

import com.google.auto.value.AutoValue;

import com.smartmobilefactory.epubreader.display.view.EpubWebView;
import com.smartmobilefactory.epubreader.utils.BaseDisposableObserver;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.subjects.BehaviorSubject;

public class VerticalEpubWebView extends EpubWebView {

    @AutoValue
    public static abstract class ScrollState {
        public abstract int top();
        public abstract int maxTop();
    }

    private BehaviorSubject<ScrollState> currentScrollState = BehaviorSubject.createDefault(new AutoValue_VerticalEpubWebView_ScrollState(0, 100));

    public VerticalEpubWebView(Context context) {
        super(context);
        init();
    }

    public VerticalEpubWebView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public VerticalEpubWebView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        verticalScrollState()
                .sample(200, TimeUnit.MILLISECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext(scrollState -> {
                    if (getProgress() == 100) {
                        callJavascriptMethod("updateFirstVisibleElement");
                    }
                })
                .subscribe(new BaseDisposableObserver<>());
    }

    @Override
    protected void onScrollChanged(int l, int t, int oldl, int oldt) {

        int height =
                (int) Math.floor(getContentHeight() * getScale());
        int webViewHeight = getMeasuredHeight();

        currentScrollState.onNext(new AutoValue_VerticalEpubWebView_ScrollState(t, height - webViewHeight));
        super.onScrollChanged(l, t, oldl, oldt);
    }

    public Observable<ScrollState> verticalScrollState() {
        return currentScrollState;
    }
}

File: library/src/main/java/com/smartmobilefactory/epubreader/display/vertical_content/VerticalEpubWebView.kt
package com.smartmobilefactory.epubreader.display.vertical_content

import android.content.Context
import android.util.AttributeSet

import com.google.auto.value.AutoValue
import com.smartmobilefactory.epubreader.display.view.EpubWebView
import com.smartmobilefactory.epubreader.utils.BaseDisposableObserver

import java.util.concurrent.TimeUnit

import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject

internal class VerticalEpubWebView @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : EpubWebView(context, attrs, defStyleAttr) {

    private val compositeDisposable = CompositeDisposable()

    private val currentScrollState = BehaviorSubject.createDefault<ScrollState>(ScrollState(0, 100))

    data class ScrollState(
            val top: Int,
            val maxTop: Int
    )

    init {
        verticalScrollState()
                .sample(200, TimeUnit.MILLISECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext {
                    if (progress == 100) {
                        callJavascriptMethod("updateFirstVisibleElement")
                    }
                }
                .subscribeWith(BaseDisposableObserver())
                .addTo(compositeDisposable)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        compositeDisposable.clear()
    }

    override fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {

        val height = Math.floor((contentHeight * scale).toDouble()).toInt()
        val webViewHeight = measuredHeight

        currentScrollState.onNext(ScrollState(t, height - webViewHeight))
        super.onScrollChanged(l, t, oldl, oldt)
    }

    fun verticalScrollState(): Observable<ScrollState> {
        return currentScrollState
    }
}
